freighter-fs-index/src/fs_index.rs

```diff
@@ -0,0 +1,243 @@
+            Ok(())
+        })
+    }
+}
+
+#[derive(Deserialize)]
+pub struct Config {
+    pub index_path: PathBuf,
+}
+
+#[async_trait]
+impl IndexProvider for FsIndexProvider {
+    type Config = Config;
+
+    async fn get_sparse_entry(&self, crate_name: &str) -> IndexResult<Vec<CrateVersion>> {
+        self.access_crate(crate_name)?.shared().await.deserialized()
+    }
+
+    async fn confirm_existence(&self, crate_name: &str, version: &Version) -> IndexResult<bool> {
+        self.access_crate(crate_name)?.shared().await
+            .deserialized()?
+            .iter()
+            .rfind(|e| &e.vers == version)
+            .map(|e| e.yanked)
+            .ok_or(IndexError::NotFound)
+    }
+
+    async fn yank_crate(&self, crate_name: &str, version: &Version) -> IndexResult<()> {
+        self.yank_inner(crate_name, version, true).await
+    }
+
+    async fn unyank_crate(&self, crate_name: &str, version: &Version) -> IndexResult<()> {
+        self.yank_inner(crate_name, version, false).await
+    }
+
+    async fn publish(
+        &self,
+        p: &Publish,
+        checksum: &str,
+        end_step: Pin<Box<dyn Future<Output = anyhow::Result<()>> + Send>>,
+    ) -> IndexResult<CompletedPublication> {
+        let release = CrateVersion {
+            name: p.name.clone(),
+            vers: p.vers.clone(),
+            deps: p
+                .deps
+                .iter()
+                .map(|d| {
+                    let (alias, package) = if let Some(renamed) = &d.explicit_name_in_toml {
+                        (renamed.clone(), Some(d.name.clone()))
+                    } else {
+                        (d.name.clone(), None)
+                    };
+                    Dependency {
+                        name: alias,
+                        req: d.version_req.clone(),
+                        features: d.features.clone(),
+                        optional: d.optional,
+                        default_features: d.default_features,
+                        target: d.target.clone(),
+                        kind: d.kind,
+                        registry: d.registry.clone(),
+                        package,
+                    }
+                })
+                .collect(),
+            cksum: checksum.into(),
+            features: p.features.clone(),
+            yanked: false,
+            links: p.links.clone(),
+            v: 2,
+            features2: Default::default(),
+        };
+
+        let lock = self.access_crate(&release.name)?;
+        let meta = lock.exclusive().await;
+
+        match meta.deserialized() {
+            Ok(existing_releases) => {
+                if existing_releases.iter().any(|v| v.vers == release.vers) {
+                    return Err(IndexError::Conflict(format!("{}-{} aleady exists", p.name, p.vers)));
+                }
+            },
+            Err(IndexError::NotFound) => {},
+            Err(other) => return Err(other),
+        };
+
+        end_step.await?;
```

Comment:
I'd recommend adding a drop guard which clears the index entry out if end_step fails.

As currently written, we can put ourselves in an inconsistent state where the index entry exists but we failed to push to the tarball store.

Reply:
Also with the storage and index you're always going to have two generals problem. I've added undoing of storage writes in case index update fails.

Reply:
Thanks, I mistook what was happening here at first.

Also, we've generally preferred to just ignore conflicts in the tarball store, and push anyways to fix this.

Reply:
I'm not sure what you mean by conflicts in the tarball store. The index has a checksum for the tarball, so you can't allow them to get out of sync. 

crates.io treats crate versions as immutable. I don't know if Cargo would react well if an old version was allowed to be overridden (even if the index updated the checksum).